api/src/test/java/org/openmrs/api/CohortServiceTest.java

```diff
@@ -564,8 +574,194 @@ public void getMemberships_shouldGetMembershipsAsOfADate() throws ParseException
+		assertEquals(service.getCohortsContainingPatient(patient, false, date).size(), 0);
+	}
+	
+	/**
+	 * @verifies return memberships on specified date
+	 * @see CohortService#getCohortsContainingPatient(org.openmrs.Patient)
+	 */
+	@Test
+	public void getCohortsContainingPatient_shouldReturnMembershipsOnSpecifiedDate() throws Exception {
+		executeDataSet(COHORT_XML);
+		
+		Cohort cohort = service.getCohort(2);
+		
+		Patient patient = new Patient(7);
+		CohortMembership membership = new CohortMembership(patient);
+		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+		Date startDate = dateFormat.parse("2017-01-01 00:00:00");
+		membership.setStartDate(startDate);
+		cohort.addMembership(membership);
+		assertTrue(cohort.contains(patient.getPatientId()));
+		
+		Date date = dateFormat.parse("2017-01-01 00:00:00");
+		assertEquals(service.getCohortsContainingPatient(patient, false, date).size(), 1);
+	}
+
+	@Test
+	public void union_shouldContainVoidedAndExpiredMemberships() throws Exception {
+		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+		Date startDate = dateFormat.parse("2017-01-01 00:00:00");
+		Date endDate = dateFormat.parse("2017-02-01 00:00:00");
+
+		Cohort cohortOne = new Cohort(3);
+		CohortMembership membershipOne = new CohortMembership(new Patient(7), startDate);
+		membershipOne.setVoided(true);
+		membershipOne.setEndDate(endDate);
+		cohortOne.addMembership(membershipOne);
+
+		Cohort cohortTwo = new Cohort(4);
+		CohortMembership membershipTwo = new CohortMembership(new Patient(8), startDate);
+		membershipTwo.setVoided(true);
+		membershipTwo.setEndDate(endDate);
+		cohortTwo.addMembership(membershipTwo);
+
+		Cohort cohortUnion = Cohort.union(cohortOne, cohortTwo);
+		Collection<CohortMembership> unionOfMemberships = cohortUnion.getMemberships();
+		assertTrue(unionOfMemberships.stream().anyMatch(m -> m.getVoided() || m.getEndDate() != null));
+	}
+
+	@Test
+	public void subtract_shouldContainVoidedAndExpiredMemberships() throws Exception {
+		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+		Date startDate = dateFormat.parse("2017-01-01 00:00:00");
+		Date endDate = dateFormat.parse("2017-02-01 00:00:00");
+
+		Cohort cohortOne = new Cohort(3);
+		CohortMembership membershipOne = new CohortMembership(new Patient(7), startDate);
+		membershipOne.setVoided(true);
+		membershipOne.setEndDate(endDate);
+		cohortOne.addMembership(membershipOne);
+
+		Cohort cohortTwo = new Cohort(4);
+		CohortMembership membershipTwo = new CohortMembership(new Patient(8), startDate);
+		membershipTwo.setVoided(true);
+		membershipTwo.setEndDate(endDate);
+		cohortTwo.addMembership(membershipTwo);
+
+		Cohort cohortSubtract = Cohort.subtract(cohortOne, cohortTwo);
+		Collection<CohortMembership> subtractOfMemberships = cohortSubtract.getMemberships();
+		assertTrue(subtractOfMemberships.stream().anyMatch(m -> m.getVoided() || m.getEndDate() != null));
```

Comment:
I thought subtract should return members that are only in cohort one and not in two. Otherwise the test above isn't concrete enough since there is no common members.